Add new README and md files (#1904)
anoma/juvix: README.md

|<a href="https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=102404734&machine=standardLinux32gb&location=WestEurope"><img height="20pt" alt="Open the Juvix Standard Lib in Github Codespace" src="https://github.com/codespaces/badge.svg"/> </a>|

Juvix is an open-source, constantly evolving functional programming language
designed for writing privacy-preserving decentralized applications. Using Juvix, developers can write high-level programs which can be compiled to WASM directly, or through [VampIR][vampir] to circuits for private execution with [Taiga][taiga] on [Anoma][anoma] or Ethereum.

## Getting Started

To get started with Juvix, head over to the [documentation website][juvix-book]
to learn more about the language and its features. You can also find
installation instructions and tutorials to help you get started with writing
Juvix programs. You can download the [latest release][latest-release] from the
Juvix GitHub repository or use the web-based development environment, [Juvix
Github Codespace][repo-codespace], which provides a pre-configured workspace
ready to use with Juvix and the Haskell toolchain installed.

## The Juvix programming language

> Juvix provides a high degree of assurance. The Juvix compiler runs several

Reviewer: ```suggestion
Juvix is designed with a focus on safety. The Juvix compiler runs several
```

static analyses which guarantee the absence of runtime errors. Analyses
performed include termination and type checking. As a result, functional
programs, especially validity predicates, can be written with greater confidence
in their correctness.

Some language features in Juvix include:

- Haskell/Agda-like syntax with support for Unicode
- Type inference
- Parametric polymorphism
- User defined inductive data types

> - Higher-order functions

Reviewer: I think it would be good to mention that Juvix is pure, perhaps by adding "Referential transparency" to the feature list:

```suggestion
- Higher-order functions
- Referential transparency
```

written with greater confidence in their correctness.

Some language features in Juvix include:

- unicode syntax
- parametric polymorphism
- inductive and parametric data types
- higher-order functions
- implicit arguments
- holes in expressions
- axioms for non-computable terms

The Juvix module system allows developers to break down their programs into smaller, reusable modules that can be compiled separately and combined to create larger programs. These modules can be used to build libraries, which can then be documented using Juvix's built-in documentation generation tool, see for example, [the Juvix standard
library's website](https://anoma.github.io/juvix-stdlib/). For further
details, please refer to [the Juvix
book](https://anoma.github.io/juvix/) which includes our [latest
updates](https://anoma.github.io/juvix/changelog.html).

## Related projects

If you're interested in Juvix, you may also want to explore the following related projects:

<!-- prettier-ignore -->
| Project | Description |
| --- | --- |
| [GEB](https://github.com/anoma/geb) | Intermediate language for writing compilers and one of the Juvix backends. |

> | [VampIR](https://github.com/anoma/vamp-ir) | Proof-system-agnostic language for writing arithmetic circuit and one of the GEB's backend.|

Reviewer: ```suggestion
| [VampIR](https://github.com/anoma/vamp-ir) | Proof-system-agnostic language for writing arithmetic circuit and one of the GEB backends.|
```

| [Juvix Github Codespace](https://github.com/codespaces-templates/juvix) | This codespace provides a pre-configured workspace with Juvix and the Haskell toolchain installed. Everything is ready to use, so you can start developing/inspecting the Juvix compiler right away. |
| [Homebrew Juvix formula](https://github.com/anoma/homebrew-juvix) | A formula for Homebrew, a package manager for macOS and Linux, that allows you to easily install Juvix on your system. |
| [Juvix Nightly builds](https://github.com/anoma/nightly-juvix-builds) | Users can download and use these nightly builds to experiment with the latest changes to the Juvix Compiler. Nightly builds may contain new features, bug fixes, and other improvements to Juvix that are still in development and have not yet been released in an official version.|

## Contributing

> If you're interested in contributing to Juvix, please see the contributing guidelines for more information. We welcome contributions of all kinds, from bug reports and feature requests to code contributions and documentation improvements.

Reviewer: Link to the contribution guidelines?

Author: Right, I was thinking of having `CONTRIBUTING.md`.
I forgot to make it.

Author: I added a preliminary version for that file.

several static analyses which guarantee the absence of runtime errors.
Analyses performed include termination, arity, and type checking. As a
result, functional programs, especially validity predicates, can be
written with greater confidence in their correctness.

Some language features in Juvix include:

- unicode syntax
- parametric polymorphism
- inductive and parametric data types
- higher-order functions
- implicit arguments
- holes in expressions
- axioms for non-computable terms

The Juvix module system allows developers to break down their programs into smaller, reusable modules that can be compiled separately and combined to create larger programs. These modules can be used to build libraries, which can then be documented using Juvix's built-in documentation generation tool, see for example, [the Juvix standard
library's website](https://anoma.github.io/juvix-stdlib/). For further
details, please refer to [the Juvix
book](https://anoma.github.io/juvix/) which includes our [latest
updates](https://anoma.github.io/juvix/changelog.html).

## Related projects

If you're interested in Juvix, you may also want to explore the following related projects:

<!-- prettier-ignore -->

> | Project | Description |

Reviewer 2: is there a way to format md tables?

Author: Yes, I dont' follow. Are you asking because the current format bothers you, or
because of the comment above the table?

Reviewer 2: The rendered table is fine, I am asking if the markdown table can be formatted to look prettier. It's not that important

Author: I see. Yes, there's a formatter that add extra "-" to make the table nicer. I'll adress that in another PR, for this and other files.

</a>

<a href="https://github.com/anoma/juvix">
<img align="right" width="270" alt="Juvix Mascot" src="assets/images/tara-seating.svg" />
</a>

Juvix is an open-source, constantly evolving functional programming language
designed for writing privacy-preserving decentralized applications. It enables
developers to write smart contracts and decentralized applications on blockchain
platforms in a secure and efficient way.

## Getting Started

To get started with Juvix, head over to the [documentation website](https://docs.juvix.org) to learn more about the language and its features. You can also find installation instructions and tutorials to help you get started with writing Juvix programs.
You can download the [latest release](https://github.com/Juvix-lang/juvix/releases) from the Juvix GitHub repository or use the web-based development environment, [Juvix Github
Codespace](https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=102404734&machine=standardLinux32gb&location=WestEurope), which provides a pre-configured workspace with Juvix and the Haskell toolchain installed and ready to use.

## The Juvix programming language

> Juvix provides a high degree of assurance. The Juvix compiler runs

Reviewer 2: This hasn't been changed, did you forget or do you prefer not to change it?

> Juvix provides a high degree of assurance. The Juvix compiler runs

Author: I prefer not to change it.

> Juvix provides a high degree of assurance. The Juvix compiler runs

Author: Thanks for the other revisions; I think I addressed them all.

several static analyses which guarantee the absence of runtime errors.

> Analyses performed include termination, arity, and type checking. As a

Reviewer 2: I would remove arity. That's included in typechecking

Author: Done.

result, functional programs, especially validity predicates, can be
written with greater confidence in their correctness.

Some language features in Juvix include:

> - unicode syntax

Reviewer 2: instead of "unicode syntax", what about "Haskell-like syntax with unicode support"

- parametric polymorphism

> - inductive and parametric data types

Reviewer 2: since the last item already mentions that we support parametric polymorphism, I would simply say "user defined inductive data types"

- higher-order functions

> - implicit arguments

Reviewer 2: While we have support for implicit arguments syntax, maybe it would better reflect the current state of the language to say "type inference". I would probably remove "holes in expressions" too, since these can only be types and that is included in type inference.

Author: Done.

- holes in expressions

> - axioms for non-computable terms

Reviewer 2: It is not clear what this means, since juvix is not (yet) a proof assistant. I would probably remove this item for now.

Author: Done.

# Juvix

<a href="https://github.com/anoma/juvix/actions/workflows/ci.yml">
<img alt="CI status" src="https://github.com/anoma/juvix/actions/workflows/ci.yml/badge.svg" />
</a>

<a href="https://github.com/anoma/juvix/actions/workflows/pages/pages-build-deployment"><img
src="https://github.com/anoma/juvix/actions/workflows/pages/pages-build-deployment/badge.svg"
alt="pages-build-deployment" /></a>

<a href="https://github.com/anoma/juvix/tags">
<img alt="" src="https://img.shields.io/github/v/release/anoma/juvix?include_prereleases" />
</a>

<a href="https://github.com/anoma/juvix/blob/main/LICENSE">
<img alt="LICENSE" src="https://img.shields.io/badge/license-GPL--3.0--only-blue.svg" />
</a>

<a href="https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=455254004">
<img height="20pt" alt="Open the Juvix Standard Lib in Github Codespace" src="https://github.com/codespaces/badge.svg" />
</a>

<a href="https://github.com/anoma/juvix">
<img align="right" width="270" alt="Juvix Mascot" src="assets/images/tara-seating.svg" />
</a>

> Juvix is an open-source, constantly evolving functional programming language
> designed for writing privacy-preserving decentralized applications. It enables
> developers to write smart contracts and decentralized applications on blockchain
> platforms in a secure and efficient way.

Reviewer 3: ```suggestion
Juvix is an open-source, constantly evolving functional programming language
designed for writing privacy-preserving decentralized applications. Using Juvix, developers can write high-level programs which can be compiled to WASM directly, or through [VampIR](https://github.com/anoma/vamp-ir) to circuits for private execution with [Taiga](https://github.com/anoma/taiga) on Anoma or Ethereum.
```

Reviewer 3: (and later we can add examples of these two cases)
